File: frontend/src/api/cashRegister.ts
import { apiFetch, type SessionLike } from "./client";

export type CashSessionStatus = "open" | "closed" | "pending_approval";
export type CashMovementType = "income" | "expense" | "adjustment";

export type CashSession = {
  id: number;
  hotel_id: number;
  opened_by_user_id?: number | null;
  closed_by_user_id?: number | null;
  status: CashSessionStatus;
  opening_balance: number;
  currency_code: string;
  opened_at: string;
  closed_at?: string | null;
  notes?: string | null;
};

export type CashMovement = {
  id: number;
  hotel_id: number;
  session_id: number;
  reservation_id?: number | null;
  transaction_id?: number | null;
  recorded_by_user_id?: number | null;
  movement_type: CashMovementType;
  amount: number;
  description?: string | null;
  recorded_at: string;
};

export type CashCloseReport = {
  id: number;
  hotel_id: number;
  session_id: number;
  closed_by_user_id?: number | null;
  expected_balance: number;
  declared_balance: number;
  difference: number;
  difference_approved: boolean;
  approved_by_user_id?: number | null;
  successor_session_id?: number | null;
  custody_handoff?: CashCustodyHandoff | null;
  notes?: string | null;
  closed_at: string;
};

export type CashCustodyHandoff = {
  id: number;
  hotel_id: number;
  close_report_id: number;
  delivered_by_user_id?: number | null;
  received_by_user_id?: number | null;
  delivered_amount: number | string;
  status: "pending" | "confirmed";
  delivered_at: string;
  received_at?: string | null;
  notes?: string | null;
};

export type CashSessionSummary = {
  session_id: number;
  status: CashSessionStatus;
  currency_code: string;
  opening_balance: number;
  income_total: number;
  expense_total: number;
  adjustment_total: number;
  confirmed_cash_total: number;
  expected_balance: number;
  movements_count: number;
};

export type CashDailyPaymentMethod = {
  payment_method: string;
  gross_collected: number;
  refunds: number;
  net_collected: number;
  transaction_count: number;
};

export type CashDailyCollector = {
  collector_user_id?: number | null;
  collector_name: string;
  gross_collected: number;
  refunds: number;
  net_collected: number;
  transaction_count: number;
};

export type CashDailyEntry = {
  entry_type: "payment" | "manual_movement";
  actor_user_id?: number | null;
  actor_name: string;
  transaction_id?: number | null;
  cash_movement_id?: number | null;
  reservation_id?: number | null;
  amount: number;
  signed_amount: number;
  currency_code: string;
  payment_method?: string | null;
  transaction_type?: string | null;
  transaction_status?: string | null;
  movement_type?: string | null;
  occurred_at: string;
  description?: string | null;
  provider_code?: string | null;
};

export type CashDailySession = {
  session_id: number;
  status: CashSessionStatus;
  currency_code: string;
  opened_at: string;
  closed_at?: string | null;
  opened_by_user_id?: number | null;
  closed_by_user_id?: number | null;
  opening_balance: number;
  expected_balance: number;
  declared_balance?: number | null;
  difference?: number | null;
};

export type CashDailySummary = {
  hotel_id: number;
  report_date: string;
  timezone: string;
  currency_code: string;
  gross_collected: number;
  refunds: number;
  net_collected: number;
  physical_cash_net_collected: number;
  digital_net_collected: number;
  by_payment_method: CashDailyPaymentMethod[];
  by_collector: CashDailyCollector[];
  physical_cash: {
    opening_balance: number;
    income_total: number;
    expense_total: number;
    adjustment_total: number;
    expected_balance: number;
    declared_balance?: number | null;
    difference?: number | null;
    manual_income_total: number;
    manual_expense_total: number;
  };
  sessions: CashDailySession[];
  entries: CashDailyEntry[];
  entries_truncated: boolean;
  generated_at: string;
};

export type CashSessionOpenPayload = {
  opening_balance: number;
  currency_code?: string;
  notes?: string | null;
};

export type CashMovementPayload = {
  movement_type: CashMovementType;
  amount: number;
  description?: string | null;
  reservation_id?: number | null;
  transaction_id?: number | null;
};

export type CashSessionClosePayload = {
  counted_balance: number;
  notes?: string | null;
  approve_difference?: boolean;
};

export const listCashSessions = (session?: SessionLike) =>
  apiFetch<CashSession[]>("/api/cash-register/sessions", { session });

export const getLatestCashCloseReport = (session?: SessionLike) =>
  apiFetch<CashCloseReport | null>("/api/cash-register/close-reports/latest", { session });

export const openCashSession = (payload: CashSessionOpenPayload, session?: SessionLike) =>
  apiFetch<CashSession>("/api/cash-register/sessions", { method: "POST", data: payload, session });

export const listCashMovements = (sessionId: number, session?: SessionLike) =>
  apiFetch<CashMovement[]>(`/api/cash-register/sessions/${sessionId}/movements`, { session });

export const getCashSessionSummary = (sessionId: number, session?: SessionLike) =>
  apiFetch<CashSessionSummary>(`/api/cash-register/sessions/${sessionId}/summary`, { session });

export const getCashDailySummary = (date: string, session?: SessionLike) =>
  apiFetch<CashDailySummary>(`/api/cash-register/daily-summary?date=${encodeURIComponent(date)}`, { session });

export const addCashMovement = (sessionId: number, payload: CashMovementPayload, session?: SessionLike) =>
  apiFetch<CashMovement>(`/api/cash-register/sessions/${sessionId}/movements`, {
    method: "POST",
    data: payload,
    session
  });

export const closeCashSession = (sessionId: number, payload: CashSessionClosePayload, session?: SessionLike) =>
  apiFetch<CashCloseReport>(`/api/cash-register/sessions/${sessionId}/close`, {
    method: "POST",
    data: payload,
    session
  });

export const approveCashCloseDifference = (reportId: number, session?: SessionLike) =>
  apiFetch<CashCloseReport>(`/api/cash-register/close-reports/${reportId}/approve`, {
    method: "POST",
    session
  });

export const confirmCashCustody = (reportId: number, session?: SessionLike) =>
  apiFetch<CashCloseReport>(`/api/cash-register/close-reports/${reportId}/custody/confirm`, {
    method: "POST",
    session
  });

File: app/schemas/cash_register.py
"""Pydantic schemas for cash register sessions, movements, and close reports."""
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, field_validator

from app.models.cash_register import CashCustodyStatusEnum, CashMovementTypeEnum, CashSessionStatusEnum


class CashSessionOpen(BaseModel):
    opening_balance: Decimal = Field(default=Decimal("0.00"), ge=Decimal("0"))
    currency_code: str = Field(default="ARS", min_length=3, max_length=3)
    notes: Optional[str] = Field(default=None, max_length=1000)

    @field_validator("currency_code")
    @classmethod
    def normalize_currency(cls, value: str) -> str:
        return value.strip().upper()


class CashSessionRead(BaseModel):
    id: int
    hotel_id: int
    opened_by_user_id: Optional[int] = None
    closed_by_user_id: Optional[int] = None
    status: CashSessionStatusEnum
    opening_balance: Decimal
    currency_code: str
    opened_at: datetime
    closed_at: Optional[datetime] = None
    notes: Optional[str] = None

    model_config = {"from_attributes": True}


class CashMovementCreate(BaseModel):
    movement_type: CashMovementTypeEnum
    amount: Decimal = Field(..., gt=Decimal("0"))
    description: Optional[str] = Field(default=None, max_length=300)
    reservation_id: Optional[int] = None
    transaction_id: Optional[int] = None


class CashMovementRead(BaseModel):
    id: int
    hotel_id: int
    session_id: int
    reservation_id: Optional[int] = None
    transaction_id: Optional[int] = None
    recorded_by_user_id: Optional[int] = None
    movement_type: CashMovementTypeEnum
    amount: Decimal
    description: Optional[str] = None
    recorded_at: datetime

    model_config = {"from_attributes": True}


class CashSessionClose(BaseModel):
    counted_balance: Decimal = Field(..., ge=Decimal("0"))
    notes: Optional[str] = Field(default=None, max_length=1000)
    approve_difference: bool = False


class CashSessionSummaryRead(BaseModel):
    session_id: int
    status: str
    currency_code: str
    opening_balance: Decimal
    income_total: Decimal
    expense_total: Decimal
    adjustment_total: Decimal
    confirmed_cash_total: Decimal
    expected_balance: Decimal
    movements_count: int


class CashDailyPaymentMethodRead(BaseModel):
    payment_method: str
    gross_collected: Decimal
    refunds: Decimal
    net_collected: Decimal
    transaction_count: int


class CashDailyCollectorRead(BaseModel):
    collector_user_id: Optional[int] = None
    collector_name: str
    gross_collected: Decimal
    refunds: Decimal
    net_collected: Decimal
    transaction_count: int


class CashDailyEntryRead(BaseModel):
    entry_type: Literal["payment", "manual_movement"]
    actor_user_id: Optional[int] = None
    actor_name: str
    transaction_id: Optional[int] = None
    cash_movement_id: Optional[int] = None
    reservation_id: Optional[int] = None
    amount: Decimal
    signed_amount: Decimal
    currency_code: str
    payment_method: Optional[str] = None
    transaction_type: Optional[str] = None
    transaction_status: Optional[str] = None
    movement_type: Optional[str] = None
    occurred_at: datetime
    description: Optional[str] = None
    provider_code: Optional[str] = None


class CashDailySessionRead(BaseModel):
    session_id: int
    status: str
    currency_code: str
    opened_at: datetime
    closed_at: Optional[datetime] = None
    opened_by_user_id: Optional[int] = None
    closed_by_user_id: Optional[int] = None
    opening_balance: Decimal
    expected_balance: Decimal
    declared_balance: Optional[Decimal] = None
    difference: Optional[Decimal] = None


class CashDailyPhysicalRead(BaseModel):
    opening_balance: Decimal
    income_total: Decimal
    expense_total: Decimal
    adjustment_total: Decimal
    expected_balance: Decimal
    declared_balance: Optional[Decimal] = None
    difference: Optional[Decimal] = None
    manual_income_total: Decimal
    manual_expense_total: Decimal


class CashDailySummaryRead(BaseModel):
    hotel_id: int
    report_date: str
    timezone: str
    currency_code: str
    gross_collected: Decimal
    refunds: Decimal
    net_collected: Decimal
    physical_cash_net_collected: Decimal
    digital_net_collected: Decimal
    by_payment_method: list[CashDailyPaymentMethodRead] = Field(default_factory=list)
    by_collector: list[CashDailyCollectorRead] = Field(default_factory=list)
    physical_cash: CashDailyPhysicalRead
    sessions: list[CashDailySessionRead] = Field(default_factory=list)
    entries: list[CashDailyEntryRead] = Field(default_factory=list)
    entries_truncated: bool = False
    generated_at: datetime


class CashCustodyHandoffRead(BaseModel):
    id: int
    hotel_id: int
    close_report_id: int
    delivered_by_user_id: Optional[int] = None
    received_by_user_id: Optional[int] = None
    delivered_amount: Decimal
    status: CashCustodyStatusEnum
    delivered_at: datetime
    received_at: Optional[datetime] = None
    notes: Optional[str] = None

    model_config = {"from_attributes": True}


class CashCloseReportRead(BaseModel):
    id: int
    hotel_id: int
    session_id: int
    closed_by_user_id: Optional[int] = None
    expected_balance: Decimal
    declared_balance: Decimal
    difference: Decimal
    difference_approved: bool
    approved_by_user_id: Optional[int] = None
    successor_session_id: Optional[int] = None
    custody_handoff: Optional[CashCustodyHandoffRead] = None
    notes: Optional[str] = None
    closed_at: datetime

    model_config = {"from_attributes": True}
